src/stream.rs

```diff
@@ -138,8 +145,11 @@ static STREAM_CALLBACK_ENVIRONMENT: UserDataKey<UnsafeCell<CallbackEnvironment>>
 struct CallbackEnvironment {
     stream: Option<Box<dyn Any>>,
     io_error: Option<io::Error>,
+    unwind_payload: Option<Box<dyn Any + Send + 'static>>,
 }
 
+// Safety: unwinding into C is undefined behavior (https://github.com/rust-lang/rust/issues/58794)
+// so code outside of the `catch_unwind` call must never panic.
```

Comment:
Isn't this actually defined behaviour now (panic) since one of the last few releases? But your solution is nicer of course.

We depend on this being defined behaviour in the other crates. Started to do so in 1.24, then it was backed out again for 1.25, and we were told that it was going to be stabilized again in some 1.3x release I can't remember right now. All this back and forth about this behaviour is really annoying.

Reply:
As the linked issue https://github.com/rust-lang/rust/issues/58794 explains, abort-on-unwind-through-`extern fn` was landed then reverted. Twice. But even if it lands again, isn’t propagating the panic preferable to aborting the process?

Reply:
In this specific case here we can do that, in general we unfortunately can't (there might be no well known place from which we would propagate the panic further) so aborting the process is the best we can do.

Not sure if we want to, yet another time, revert back to adding guards to the other crates to exit the process at the FFI boundary. It's getting boring and causes a lot of churn, and the undefined behaviour right now seems to be to abort the process...

Not complaining to you personally here btw :) Thanks for providing the link to the latest version of the adventure.

Reply:
> there might be no well known place from which we would propagate the panic further

Indeed, https://github.com/gtk-rs/cairo/issues/262 is such a case.

The PDF / PS / SVG output stream case is sort of in-between. And we’re not alone with these questions, the end of https://www.cairographics.org/manual/bindings-streams.html discusses it as well.